[release/8.0.4xx] Multi-arch OCI images export as tarballs

surayya-MS opened the pull request:
### Context

Fixes https://github.com/dotnet/sdk-container-builds/issues/609

This is a feature request from partner team - support publishing multi-arch OCI image as one combined tarball. Currently we publish separate tarballs for each architecture.

### Customer impact

Without this feature customers will not be able to export multi-arch OCI image as one combined tarball, they will get separate tarballs for each architecture, and will have to combine them manually to achieve this.

### Details

**Current behavior for multi-arch image publishing**

- remote registry - create single-arch image for each arch and publish them to the remote registry, then create image index based on them and load to remote registry
- local daemon (docker/podman) - create single-arch image for each arch and publish them to the local daemon
- tarball - create single-arch image for each arch and publish them as separate tarballs

**Behaviour after this PR**

- remote registry - create single-arch image for each arch and publish them to the remote registry, then create image index based on them and load to remote registry (same as before)
- local daemon (docker) - create one multi-arch image and publish it to the local daemon
- local daemon (podman) - create single-arch image for each arch and publish them to the local daemon (same as before; support for podman will be added in the future)
- tarball - create one multi-arch image and publish it as a tarball

_Note_: in order to load a multi-arch oci image tarball, containerd image store should be enabled for Docker.

### Changes made 

1. Added support for multi-arch OCI tarballs creation in `DockerCli`.
2. Refactored the code to extract publishing logic from `CreateNewImage` task into a separate class, to be able to reuse it in `CreateImageIndex` task.
3. Added `SkipPublishing` parameter in `CreateNewImage` task, in order to create an image but skip publish it (this is needed for multi-arch tarball publishing because we need to publish one combined tarball)
4. More refactoring

### Testing
Modified end to end multi-arch tests to reflect the new behavior. Tested manually as well.

### Risks
_Low_  - new tests ensure the scenarios are now covered, the existing tests ensure that other scenarios were not affected.

surayya-MS committed:
check docker availability before arch support in tests

Commit:
fix e2e test after checking option 'Use containerd for pulling and storing images'

Commit:
refactor BuiltImage

Commit:
refactor single-arch oci tarball publishing

Commit:
implement loading multi-arch image as a tarball and to a local registry

Commit:
implement publishing to tarball/local daemon/remote registry in CreateImageIndex

Commit:
move Telemetry into a seperate file

Commit:
move Telemetry class to another folder

Commit:
extract publishing image logic into a seperate class

Commit:
refactor ImagePublisher

Commit:
skip publishing individual images in case of tarball and local daemon docker publishing

Commit:
update public api

Commit:
skip CreateImageIndex task in case of local damon podman publishing as it is not supported

Commit:
refactor CreateImageIndex task; improve logging; log tip for enabling conrainerd

Commit:
add support for multi tagoci tarball

Commit:
fix test EndToEndMultiArch_LocalRegistry

Commit:
refactoring

Commit:
fixed multi-arch e2e tests

Commit:
cleanup

Commit:
refactoring: make BuiltImage a class instead of a struct

Commit:
rearrange functions order in DockerCli for easier review

Commit:
rearrange functions again

Commit:
add back log message about building image index

Commit:
extract error messages into Strings

baronfel commented on src/Containers/Microsoft.NET.Build.Containers/Resources/Strings.resx:
```diff
@@ -358,26 +355,54 @@
+    <comment/>
+  </data>
+  <data name="TipToEnableContainerdForMultiArch" xml:space="preserve">
+    <value>Tip: For multi-arch image publishing, ensude that 'Use containerd for pulling and storing images' is checked in Docker Desktop settings.</value>
```
```suggestion
    <value>Tip: For multi-arch image publishing, ensure that 'Use containerd for pulling and storing images' is checked in Docker Desktop settings.</value>
```

surayya-MS replied:
thanks! i later decided to change this. instead of guessing that the containrd image store is disabled based on the error from docker, i now check if it is enabled before loading the image

Commit:
fix typo in log message

Commit:
extract image index creation from DockerCli into ImageIndexGenerator

Commit:
cleanup

Commit:
add more ImageIndexGenertorTests

Commit:
check if containerd store is enabled before loading

surayya-MS commented:
`EndToEndMultiArch_ContainerRuntimeIdentifiersOverridesRuntimeIdentifiers` and `EndToEndMultiArch_Labels` tests are failing in `dotnet-sdk-public-ci (Build Ubuntu_22_04 Build_Release)` because containerd image store is not enabled for docker.
Other multi-arch containers tests are not failing because they are skipped for `arm64` architecture.

@baronfel do you know who I can ask to enable containerd image store?

surayya-MS commented:
PR is ready for review

baronfel commented:
> `EndToEndMultiArch_ContainerRuntimeIdentifiersOverridesRuntimeIdentifiers` and `EndToEndMultiArch_Labels` tests are failing in `dotnet-sdk-public-ci (Build Ubuntu_22_04 Build_Release)` because containerd image store is not enabled for docker. Other multi-arch containers tests are not failing because they are skipped for `arm64` architecture.
> 
> Also, `dotnet-sdk-public-ci (Build Darwin Build_Release)` check fails on macos - this unrelated to this PR
> 
> @baronfel, @rainersigwald do you know who I can ask to enable containerd image store?

@rainersigwald pointed me to dnceng and the machine definitions. They seem to have containerd already installed, but not enabled. Will start a thread and pull you into it.

baronfel commented:
Ok, suggestion is to:

* write the containerd config to the daemon in the build pipeline
* restart the docker daemon to pick up the config change
* see if that unblocks us

To write the config, we'd need to add the following json to `/etc/docker/daemon.json`:

```json
{
  "features": {
    "containerd-snapshotter": true
  }
}
```

To restart the daemon we should just have to run ` sudo systemctl restart docker`.

JanKrivanek reviewed (commented):
I do not have strong concerns - but it's a big PR - I'll need a second deeper look

baronfel commented:
I've been playing with this locally and saw a couple behaviors that I wanted to note here.

# Local Docker behaviors

* when published, the multi-arch image generates two images with the same Image Id, size, etc. This feels incorrect to me. I can't `docker run` these images from the command line, and `docker image inspect` only works if I use the full digest SHA. This is a UX regression that we must fix.

```terminal
>docker image list sdk-container-demo
REPOSITORY           TAG                  IMAGE ID       CREATED         SIZE
sdk-container-demo   latest               18d9c080eb5f   8 minutes ago   466MB
sdk-container-demo   latest               18d9c080eb5f   8 minutes ago   466MB
```

![image](https://github.com/user-attachments/assets/35e687ce-c5c3-4e87-a027-98893855b800)

```terminal
> docker run -it --rm sdk-container-demo:latest
Unable to find image 'sdk-container-demo:latest' locally
docker: Error response from daemon: pull access denied for sdk-container-demo, repository does not exist or may require 'docker login'.
See 'docker run --help'.
```

# Tarball export

Setting `ContainerArchiveOutputPath` to a relative directory that does not exist results in a tarball written to the relative path, but without the `.tar.gz` extension:

```terminal
> dotnet build -t:PublishContainer -p RuntimeIdentifiers=`"linux-x64`;linux-arm64`" -p ContainerArchiveOutputPath="./container-tarball" -bl -v d -tl
Restore complete (0.6s)
    Determining projects to restore...
    All projects are up-to-date for restore.
  sdk-container-demo succeeded (2.1s) → E:\Code\sdk-container-demo\artifacts\publish\sdk-container-demo\debug_linux-arm64\
    Building image 'sdk-container-demo' for runtime identifier 'linux-arm64' on top of base image 'mcr.microsoft.com/dotnet/aspnet:8.0'.
  sdk-container-demo succeeded (2.5s) → E:\Code\sdk-container-demo\artifacts\publish\sdk-container-demo\debug_linux-x64\
    Building image 'sdk-container-demo' for runtime identifier 'linux-x64' on top of base image 'mcr.microsoft.com/dotnet/aspnet:8.0'.
  sdk-container-demo succeeded (0.0s)
    Building image index 'sdk-container-demo:latest' on top of manifests sha256:401091fa8c61a10c9b621601e279143cca3b39158cce2eb5808dda5374a88a73, sha256:e29c20608fafef24e5a2c55e4dec5e9d1842a0c42641d857dceb2d9ea0b2c040.
    Pushed image 'sdk-container-demo:latest' to local archive at 'E:\Code\sdk-container-demo\src\sdk-container-demo\container-tarball'.

Build succeeded in 4.0s
```

Trying to specify a path with a .tar.gz extension results in an error:
```terminal
>dotnet build -t:PublishContainer -p RuntimeIdentifiers=`"linux-x64`;linux-arm64`" -p ContainerArchiveOutputPath="./container-tarball.tar.gz" -bl -v d -tl
Restore complete (0.7s)
    Determining projects to restore...
    All projects are up-to-date for restore.
  sdk-container-demo failed with 1 error(s) (0.0s)
    E:\Code\sdk-8.0.4xx\artifacts\bin\redist\Debug\dotnet\sdk\8.0.406-dev\Containers\build\Microsoft.NET.Build.Containers.targets(300,5): error : ContainerArchiveOutputPath is invalid. It can only be a directory in multi-arch scenario because the runtime identifier is appended in the end of the filename.

Build failed with 1 error(s) in 1.1s
```
We should clean this up - specifying a .tar.gz when doing a multi-arch tarball export shouldn't fail, and if a path is specified and it doesn't exist we should create a tar.gz extension.

For all of this, the behavior when pushing multiple single-arch tarballs should be preserved still.

baronfel commented on src/Containers/Microsoft.NET.Build.Containers/LocalDaemons/DockerCli.cs:
```diff
@@ -540,6 +625,37 @@ private static bool IsPodmanAlias()
         }
     }
 
+    private static bool IsContainerdStoreEnabledForDocker()
```
This is not a typo, funnily enough - the tech we're checking here is called `containerd`.

JanKrivanek replied:
:-O

baronfel commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs:
```diff
@@ -11,49 +12,8 @@
-    /// The generated image index (manifest list) in JSON format.
-    /// </summary>
-    [Output]
-    public string GeneratedImageIndex { get; set; }
```
For debugging purposes, it would be nice to have the output from the task. It would be good to have the digest and the content so we can figure out what went where.

surayya-MS replied:
thanks! done

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/ImagePublisher.cs:
```diff
@@ -0,0 +1,197 @@
+        if (!(await localRegistry.IsAvailableAsync(cancellationToken).ConfigureAwait(false)))
+        {
+            telemetry.LogMissingLocalBinary();
+            Log.LogErrorWithCodeFromResources(nameof(Strings.LocalRegistryNotAvailable));
```
Great question! It is a result of my refactoring - I removed SafeLog function (it checked if BuildEngine is not null before logging) because it had only 2 calls. I don't know why initially we need to check BuildEngine. I will investigate  I need to check if BuildEngine is null that other log statements.

JanKrivanek replied:
Feel free to resolve here yourself once done

surayya-MS replied:
@rainersigwald do you think we should leave BuildEngine null checks as is or it is safe to remove them?

surayya-MS replied:
Thanks! I checked once more in [`CreateNewImage` task](https://github.com/dotnet/sdk/blob/release/8.0.4xx/src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateNewImage.cs) (where I extracted image publishing from) and a lot of places logging message or error is called without checking BuildEngine is null or not. It was probably there to avoid exception for some test where the task was created with no BuildEngine. I think it is safe to remove.

surayya-MS commented on src/Containers/packaging/build/Microsoft.NET.Build.Containers.targets:
```diff
@@ -302,6 +303,18 @@
+      <!--We don't want to skip publishing individual images in case of remote registry because the individual images should be available in the registry before image index is pushed.-->
+      <!--We don't want to skip publishing individual images in case of local daemon podman because podman loads multi-arch tarball differently - only individual image for the current platform.-->
+      <_SkipContainerPublishing>false</_SkipContainerPublishing>
+      <_SkipContainerPublishing Condition="$(ContainerArchiveOutputPath) != '' or ( $(ContainerRegistry) == '' and ( $(LocalRegistry) == '' or $(LocalRegistry) == 'Docker' ) )">true</_SkipContainerPublishing>
```
I thought about this case - customer setting `_SkipContainerPublishing` to true which would lead to incorrect `PublishContainer` target behaviour. 
The solution is to not let to overwrite it. If a customer sets `_SkipContainerPublishing` to true, it will be reset it to false on line 310, and then on line 311 see if we really need to skip it or not.
Same with `_SkipCreateImageIndex` - it cannot be overwritten by a customer.

surayya-MS replied:
I just checked and in single-arch scenario you can set `_SkipContainerPublishing=true` to skip container publishing. I will change it, so it would stay consistent - not able to overwrite in both single-arch and multi-arch cases

surayya-MS commented:
> # Local Docker behaviors
> * when published, the multi-arch image generates two images with the same Image Id, size, etc. This feels incorrect to me. I can't `docker run` these images from the command line, and `docker image inspect` only works if I use the full digest SHA. This is a UX regression that we must fix. It feels like maybe we have the wrong IDs?
> 
> ```
> >docker image list sdk-container-demo
> REPOSITORY           TAG                  IMAGE ID       CREATED         SIZE
> sdk-container-demo   latest               18d9c080eb5f   8 minutes ago   466MB
> sdk-container-demo   latest               18d9c080eb5f   8 minutes ago   466MB
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/573979/409277667-35e687ce-c5c3-4e87-a027-98893855b800.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg2NzgyOTksIm5iZiI6MTczODY3Nzk5OSwicGF0aCI6Ii81NzM5NzkvNDA5Mjc3NjY3LTM1ZTY4N2NlLWM1YzMtNGU4Ny1hMDI3LTk4ODkzODU1YjgwMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNFQxNDA2MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MThlZTA2YjlmYTYxYmJmYTk3YjAwMDNmYTRiODUwZDYwYmE1YmY0N2ZiMmNmZGFmNmM4NzBmMzY0ZWJhM2Y0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4-syPA8QY8XKBV3Vaxu1ry_slVPXaVGf7JcWihmVT5M)
> 
> ```
> > docker run -it --rm sdk-container-demo:latest
> Unable to find image 'sdk-container-demo:latest' locally
> docker: Error response from daemon: pull access denied for sdk-container-demo, repository does not exist or may require 'docker login'.
> See 'docker run --help'.
> ```
> 
> Judging from [moby/moby#48902](https://github.com/moby/moby/pull/48902) and [moby/moby#48759](https://github.com/moby/moby/issues/48759), save and load don't _really_ support multi-arch manifests at this time. I wonder if that's enough to pull the feature back and just allow creating the single-arch images when directly loading into Docker at this time.

@baronfel It might be a docker bug, but I'm not 100% sure. Docker doesn't support loading multi-arch oci tarballs unless containerd image store is enabled. And when it is , the behavior is as you described. I remember when Brandyn was showing us how he loads the multi-arch tarball he mentioned that docker shows it as 2 images instead of 1. I however, don't remember if he was running the containers with image id or image name.

I previously investigated the contents of the tarball to make sure there are no duplications anywhere that might cause this, and didn't find any. The only case when docker shows the loaded oci image as 1 and not 2, is when name annotations are not specified - "io.containerd.image.name" and "org.opencontainers.image.ref.name". I tried to set only one of them at a time and the result was docker showing 2 images.

Le't wait for the partner team to confirm this.

surayya-MS commented:
> # Tarball export
> Setting `ContainerArchiveOutputPath` to a relative directory that does not exist results in a tarball written to the relative path, but without the `.tar.gz` extension:
> 
> ```
> > dotnet build -t:PublishContainer -p RuntimeIdentifiers=`"linux-x64`;linux-arm64`" -p ContainerArchiveOutputPath="./container-tarball" -bl -v d -tl
> Restore complete (0.6s)
>     Determining projects to restore...
>     All projects are up-to-date for restore.
>   sdk-container-demo succeeded (2.1s) → E:\Code\sdk-container-demo\artifacts\publish\sdk-container-demo\debug_linux-arm64\
>     Building image 'sdk-container-demo' for runtime identifier 'linux-arm64' on top of base image 'mcr.microsoft.com/dotnet/aspnet:8.0'.
>   sdk-container-demo succeeded (2.5s) → E:\Code\sdk-container-demo\artifacts\publish\sdk-container-demo\debug_linux-x64\
>     Building image 'sdk-container-demo' for runtime identifier 'linux-x64' on top of base image 'mcr.microsoft.com/dotnet/aspnet:8.0'.
>   sdk-container-demo succeeded (0.0s)
>     Building image index 'sdk-container-demo:latest' on top of manifests sha256:401091fa8c61a10c9b621601e279143cca3b39158cce2eb5808dda5374a88a73, sha256:e29c20608fafef24e5a2c55e4dec5e9d1842a0c42641d857dceb2d9ea0b2c040.
>     Pushed image 'sdk-container-demo:latest' to local archive at 'E:\Code\sdk-container-demo\src\sdk-container-demo\container-tarball'.
> 
> Build succeeded in 4.0s
> ```
> 
> Trying to specify a path with a .tar.gz extension results in an error:
> 
> ```
> >dotnet build -t:PublishContainer -p RuntimeIdentifiers=`"linux-x64`;linux-arm64`" -p ContainerArchiveOutputPath="./container-tarball.tar.gz" -bl -v d -tl
> Restore complete (0.7s)
>     Determining projects to restore...
>     All projects are up-to-date for restore.
>   sdk-container-demo failed with 1 error(s) (0.0s)
>     E:\Code\sdk-8.0.4xx\artifacts\bin\redist\Debug\dotnet\sdk\8.0.406-dev\Containers\build\Microsoft.NET.Build.Containers.targets(300,5): error : ContainerArchiveOutputPath is invalid. It can only be a directory in multi-arch scenario because the runtime identifier is appended in the end of the filename.
> 
> Build failed with 1 error(s) in 1.1s
> ```
> 
> We should clean this up - specifying a .tar.gz when doing a multi-arch tarball export shouldn't fail, and if a path is specified and it doesn't exist we should create a tar.gz extension.

I'll fix this

Commit:
cleanup targets

Commit:
add docker.io/library prefix to image name annotation; fix json serialization symbol escaping

Commit:
delete empty lines

Commit:
fix EndToEndMultiArch_ArchivePublishing test

Commit:
update EndToEndMultiArch_RemoteRegistry test since containerd store is enabled

JanKrivanek commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -2,99 +2,120 @@
     {
+        if (images.Length == 0)
+        {
+            throw new ArgumentException(string.Format(Strings.ImagesEmpty));
```
```suggestion
            throw new ArgumentException(Strings.ImagesEmpty);
```

No need for format (same on line 23)

surayya-MS replied:
thanks! done

JanKrivanek reviewed (approved):
Overall looks good.
But you might want to get as well review from someone with closer domain knowledge

Commit:
go back to running containers by image and not imageId for multi-arch tests

Commit:
implement skipping tests if containerd image store is not enabled

Commit:
remove using

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-tarballs

Commit:
refactor to bring back GeneratedImageIndex; fix ImageIndexGeneratorTests

Commit:
small fix

Commit:
Merge remote-tracking branch 'origin/multi-arch-tarballs' into multi-arch-tarballs

surayya-MS commented:
The changes made from the last review:

1. Fixed multi-arch oci tarball by prefixing name annotation by 'docker.io/library/', so that docker would show one image after loading the tarball.
2. Fixed setting `ContainerArchiveOutputPath` to a file with `.tar.gz` extension. 
3. Added condition to custom test attributes, to skip the test if containerd image store is not enabled. 
4. Brought back `GeneratedImageIndex` output parameter to `CreateImageIndex`

A test is failing on macos leg - which is unrelated to the changes in this PR.

The PR is ready for review.

MichalPavlik commented on src/Containers/Microsoft.NET.Build.Containers/DigestUtils.cs:
```diff
@@ -17,6 +17,16 @@ internal sealed class DigestUtils
+    {
+        if (!digest.StartsWith("sha256:", StringComparison.OrdinalIgnoreCase))
+        {
+            throw new ArgumentException("Invalid digest format. Digest must start with 'sha256:'.");
```
I saw localized exception messages in this project, so I think this should be localized too.

surayya-MS replied:
thanks for the suggestion! 
I think this is fine not localized, as the exception is not intended for the customers. this is just a check for an engineer if they misuse it.
If you or anyone else feels strongly about this please unresolve the conversation.

MichalPavlik commented on src/Containers/Microsoft.NET.Build.Containers/BuiltImage.cs:
```diff
@@ -24,28 +24,43 @@ internal readonly struct BuiltImage
         get
         {
-            List<ManifestLayer> layersNode = Manifest.Layers ?? throw new NotImplementedException("Tried to get layer information but there is no layer node?");
+            List<ManifestLayer> layersNode = Layers ?? throw new NotImplementedException("Tried to get layer information but there is no layer node?");
```
I'm not sure if this should be localized

surayya-MS replied:
same here

MichalPavlik commented on src/Containers/Microsoft.NET.Build.Containers/MultiArchImage.cs:
```diff
@@ -0,0 +1,16 @@
+/// <summary>
+/// Represents constructed image ready for further processing.
+/// </summary>
+internal sealed class MultiArchImage
```
nit: This could be a readonly struct

MichalPavlik commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs:
```diff
@@ -63,6 +23,8 @@ public void Dispose()
         _cancellationTokenSource.Dispose();
     }
 
+    private bool IsLocalPull => string.IsNullOrEmpty(BaseRegistry);
```
nit: I think that `IsNullOrWhiteSpace` would be a better fit here. Just in case someone will provide only white spaces :)

surayya-MS replied:
thanks, I haven't thought of that! done.

Commit:
make BuiltImage and MultiArchImage readonly structs

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-tarballs

Commit:
if ArchiveOutputPath does't have file extension then treat it as directory; add ArchiveFileRegistryTests

Commit:
trim BaseRegistry before checking null or empty

Commit:
BaseRegistry shouldn't be required parameter

Commit:
address comments

Commit:
refactor ImagePublisher

Commit:
remove BuildEngine null check before logging

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-tarballs

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateNewImage.cs:
```diff
@@ -25,7 +23,7 @@ public sealed partial class CreateNewImage : Microsoft.Build.Utilities.Task, ICa
     public string ToolPath { get; set; }
 
-    private bool IsLocalPull => string.IsNullOrEmpty(BaseRegistry);
+    private bool IsLocalPull => string.IsNullOrEmpty(BaseRegistry.Trim());
```
thanks! done

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/LocalDaemons/ArchiveFileRegistry.cs:
```diff
@@ -15,14 +14,22 @@ public ArchiveFileRegistry(string archiveOutputPath)
+        var directorySeparatorChar = Path.DirectorySeparatorChar;
+
+        // if doesn't end with a file extension, assume it's a directory
+        if (!fullPath.Split(directorySeparatorChar).Last().Contains('.'))
```
thanks! done

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -1,100 +1,126 @@
+        };
+        var escapeOptions = new JsonSerializerOptions
+        {
+            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping
```
thanks! done

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/DigestUtils.cs:
```diff
@@ -17,6 +17,16 @@ internal sealed class DigestUtils
+    {
+        if (!digest.StartsWith("sha256:", StringComparison.OrdinalIgnoreCase))
+        {
+            throw new ArgumentException("Invalid digest format. Digest must start with 'sha256:'.");
```
thanks! done

Commit:
register task resources for CreateImageIndex

Commit:
optimize ParseImages in CreateImageIndex

Commit:
fix setting annotation image name prefix: set to 'docker.io/' when repostory contains /, and 'docker.io/library/' when it doesn't; add tests

Commit:
move DockerIsAvailableAndSupportsArchTheoryAttribute to its own file; rename DockerIsAvailableAndSupportsArchFactAttribute file

baronfel commented:
I've done all my testing and am very happy with the changes here! Thanks for pushing this to done, @surayya-MS

surayya-MS commented:
A test is failing on macos leg - which is unrelated to the changes in this PR.

@marcpopMSFT could you please merge this on red, once tactics approve the PR? 
Also, could you please remove co-author from squash commit <surayyahuseynzada@Surayyas-MacBook-Pro.local>? I just noticed I haven't set git config correctly.

marcpopMSFT commented:
It's currently marked for do not merge and still servicing-consider. I can merge once it's approved. The Mac issue is #46006 and still waiting on a fix from Runtime.

baronfel commented:
@marcpopMSFT this was just approved over email - mind merging when you have a moment?
